File: lib/test/t_GaussianProcessRandomVector_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of GaussianProcessRandomVector class
 *
 *  Copyright 2005-2025 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    PlatformInfo::SetNumericalPrecision(2);

    // Learning data
    Point levels = {8, 5};
    // Define the Box
    Box box(levels);
    // Get the input sample
    Sample inputSample( box.generate() );
    // Scale each direction
    inputSample *= 10;

    // Define model
    Description inputDescription = {"x", "y"};

    Description formula = {"cos(0.5*x) + sin(y)"} ;
    const SymbolicFunction model(inputDescription, formula);

    // Build the output sample
    const Sample outputSample(model(inputSample));

    // 2) Definition of exponential model
    Point scale = {5.33532, 2.61534};
    Point amplitude = {1.61536};

    SquaredExponential covarianceModel(scale, amplitude);

    // 3) Basis definition
    Basis basis(ConstantBasisFactory(2).build());

    // Gaussian Process fitter
    GaussianProcessFitter algo(inputSample, outputSample, covarianceModel, basis);
    algo.setOptimizeParameters(false);
    algo.run();

    // Regression
    GaussianProcessRegression regression(algo.getResult());
    regression.run();

    // Get result
    GaussianProcessRegressionResult result(regression.getResult());

    // Get meta model
    Function metaModel(result.getMetaModel());

    // Interpolation error
    assert_almost_equal(outputSample,  metaModel(inputSample), 3.0e-5, 3.0e-5);


    // variance is 0 on learning points
    GaussianProcessConditionalCovariance gpcc(result);
    const CovarianceMatrix var(gpcc.getConditionalCovariance(inputSample));

    // assert_almost_equal could not be applied to matrices
    Point covariancePoint(*var.getImplementation());
    assert_almost_equal(covariancePoint, Point(covariancePoint.getSize()), 1e-6, 1e-6);

    // Random vector evaluation
    Sample unifRealization(Uniform(0.0, 10.0).getSample(2));
    Point validationPoint(unifRealization.getImplementation()->getData());
    GaussianProcessRandomVector rvector(result, validationPoint);

    // Realization of the random vector
    Point realization (rvector.getRealization());
    std::cout << "Realization of the GPRV=" << realization << std::endl;

    // Get a sample of size 10
    Sample realizations(rvector.getSample(10));
    std::cout << "Sample of realizations of the GPRV=" << realizations << std::endl;
  }

  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }

  return ExitCode::Success;
}

File: python/test/t_GaussianProcessRandomVector_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.experimental as otexp
import openturns.testing as ott

ot.TESTPREAMBLE()

ot.PlatformInfo.SetNumericalPrecision(3)
# Kriging use case
inputDimension = 2

# Learning data
levels = [8.0, 5.0]
box = ot.Box(levels)
inputSample = box.generate()
# Scale each direction
inputSample *= 10


model = ot.SymbolicFunction(["x", "y"], ["cos(0.5*x) + sin(y)"])
outputSample = model(inputSample)

# Validation data
sampleSize = 10
inputValidSample = ot.JointDistribution(2 * [ot.Uniform(0, 10.0)]).getSample(sampleSize)
outputValidSample = model(inputValidSample)

# 2) Definition of exponential model
# The parameters have been calibrated using TNC optimization
# and AbsoluteExponential models
covarianceModel = ot.SquaredExponential([7.63, 2.11], [7.38])

# 3) Basis definition
basis = ot.ConstantBasisFactory(inputDimension).build()

# Gaussian Process Regression algorithm
fit_algo = otexp.GaussianProcessFitter(
    inputSample, outputSample, covarianceModel, basis
)
fit_algo.setOptimizeParameters(False)  # do not optimize hyper-parameters
fit_algo.run()
gpr_algo = otexp.GaussianProcessRegression(fit_algo.getResult())
gpr_algo.run()
gpr_result = gpr_algo.getResult()

# Gaussian Process Conditional Covariance
gpcc = otexp.GaussianProcessConditionalCovariance(gpr_result)
# Get meta model
metaModel = gpr_result.getMetaModel()
outData = metaModel(inputValidSample)

# 4) Errors
# Interpolation
ott.assert_almost_equal(outputSample, metaModel(inputSample), 3.0e-5, 3.0e-5)


# 5) GP variance is 0 on learning points
var = gpcc.getConditionalCovariance(inputSample)

# assert_almost_equal could not be applied to matrices
# application to Point
covariancePoint = ot.Point(var.getImplementation())
trueVariance = ot.Point(covariancePoint.getSize(), 0.0)
ott.assert_almost_equal(covariancePoint, trueVariance, 1e-6, 1e-6)

# Random vector evaluation
rvector = otexp.GaussianProcessRandomVector(gpr_result, inputValidSample[0])

# Realization of the random vector
realization = rvector.getRealization()
print("Realization of the KRV=", realization)

# Get a sample of size 10
realizations = rvector.getSample(10)
print("Sample of realizations of the KRV=", realizations)
